go/worker/beacon/worker.go

```diff
@@ -0,0 +1,588 @@
+	return w.quitCh
+}
+
+func (w *Worker) Cleanup() {
+}
+
+func (w *Worker) Name() string {
+	return "beacon worker"
+}
+
+func (w *Worker) worker() {
+	defer func() {
+		close(w.quitCh)
+	}()
+
+	// Wait for consensus to be synced.
+	select {
+	case <-w.stopCh:
+		return
+	case <-w.consensus.Synced():
+	}
+
+	// Handle persisted state if required.
+	w.recoverPersistedState()
+
+	// Subscribe to SCRAPE events.
+	eventCh, eventSub := w.backend.WatchLatestSCRAPEEvent()
+	defer eventSub.Close()
+
+	for {
+		var ev *beacon.SCRAPEEvent
+		select {
+		case <-w.stopCh:
+			return
+		case ev = <-eventCh:
+			w.logger.Debug("beacon event",
+				"height", ev.Height,
+				"epoch", ev.Epoch,
+				"round", ev.Round,
+				"state", ev.State,
+			)
+		}
+
+		// Regardless of what the state transition is, we need to
+		// get the backend beacon state.
+		scrapeState, err := w.backend.GetSCRAPEState(w.ctx, ev.Height)
+		if err != nil {
+			w.logger.Error("failed to query SCRAPE state",
+				"err", err,
+			)
+			continue
+		}
+
+		switch ev.State {
+		case beacon.StateCommit:
+			w.onStateCommit(ev, scrapeState)
+		case beacon.StateReveal:
+			w.onStateReveal(ev, scrapeState)
+		case beacon.StateComplete:
+			w.onStateComplete(ev, scrapeState)
+		}
+	}
+}
+
+func (w *Worker) eventOkForState(ev *beacon.SCRAPEEvent, expectedState beacon.RoundState) bool {
+	if w.state == nil {
+		return false
+	}
+	if ev.Epoch != w.state.Epoch || ev.Round != w.state.Round {
+		w.logger.Error("epoch/round mismatch",
+			"epoch", w.state.Epoch,
+			"round", w.state.Round,
+		)
+		return false
+	}
+	if w.state.State != expectedState {
+		w.logger.Error("round state mismatch",
+			"state", w.state.State,
+		)
+		return false
+	}
+
+	return true
+}
+
+func (w *Worker) onStateCommit(ev *beacon.SCRAPEEvent, scrapeState *beacon.SCRAPEState) {
+	if w.state != nil {
+		if ev.Epoch == w.state.Epoch && ev.Round == w.state.Round {
+			// This shouldn't happen, log something about it.
+			w.logger.Error("commit: received redundant round event, ignoring")
+			return
+		}
+		w.persistState(nil)
+	}
+
+	// New round.
+
+	// Determine if we are present in the participant list, and ensure
+	// that the point for each participant are valid.
+	var (
+		isPresent bool
+		points    []scrape.Point
+	)
+	for _, id := range ev.Participants {
+		if id.Equal(w.identity.NodeSigner.Public()) {
+			isPresent = true
+		}
+		node, err := w.registry.GetNode(
+			w.ctx,
+			&registry.IDQuery{
+				Height: ev.Height,
+				ID:     id,
+			},
+		)
+		if err != nil {
+			w.logger.Error("commit: failed to query node descriptor",
+				"err", err,
+				"id", id,
+			)
+			return
+		}
+		if node.Beacon == nil {
+			w.logger.Error("commit: participant node missing point",
+				"id", id,
+			)
+			return
+		}
+		points = append(points, node.Beacon.Point)
+	}
+	h, cmpH := hash.NewFrom(points), hash.NewFrom(scrapeState.Instance.Participants)
+	if !h.Equal(&cmpH) {
+		w.logger.Error("commit: participant point list mismatch")
```

Comment:
does anything use the event that doesn't already know the correct participant point list?

Reply:
What do you mean?  In theory the registry contains the points for each node, and in theory the consensus layer isn't producing busted events, so a simple search would suffice I suppose.  The check is cheap (the number of participants is unlikely to exceed 20 or so)...